### PR TITLE
Fix assembling of xchg and bswap

### PR DESCRIPTION
**Pull request by atodekangae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

Previously, rasm2 incorrectly assembled `bswap` and `xchg` for x86-64 as shown below.

```
$ rasm2 -b 64 -
bswap r15
480fcf
bswap r15d
0fcf
xchg al, dl
92
xchg dl, al
92
xchg ax, dx
92
xchg rax, rdx
92
xchg rdx, rbx
87da
xchg rbx, rdx
87d3
xchg r8, r15
97
xchg r15, r8
97
xchg r8d, r15d
97
xchg r15d, r8d
97
```

While the expected behavior is as follows.

```
$ rasm2 -b 64 -
bswap r15
490fcf
bswap r15d
410fcf
xchg al, dl
86c2
xchg dl, al
86d0
xchg ax, dx
6692
xchg rax, rdx
4892
xchg rdx, rbx
4887d3
xchg rbx, rdx
4887da
xchg r8, r15
4d87c7
xchg r15, r8
4d87f8
xchg r8d, r15d
4587c7
xchg r15d, r8d
4587f8
```

This pull request fixes this problem.

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: test/db/asm/x86_32 [hunk 1]
```diff
@@ -2156,9 +2162,9 @@ a "xchg eax, edi" 97
 a "xchg eax, edx" 92
 a "xchg eax, esi" 96
 a "xchg eax, esp" 94
-a "xchg ebx, ecx" 87cb
-a "xchg ecx, ebp" 87e9
-a "xchg ecx, ebx" 87d9
+a "xchg ebx, ecx" 87d9
+a "xchg ecx, ebp" 87cd
+a "xchg ecx, ebx" 87cb
```
Comment (atodekangae) [inline test/db/asm/x86_32:2167]: Although the original output was not incorrect per Intel's Instruction Set Reference (`xchg r32, r32` can be interpreted either as an instance of `xchg r32, r/m32` or as `xchg r/m32, r32`), I fixed these lines and the corresponding code to make it in line with the output of other assemblers (NASM and FASM).

Comment (ret2libc) [reply test/db/asm/x86_32:2167]: Mmmm they do seem like different instructions to me. Why the change? Wasn't it working already before?

Comment (atodekangae) [reply test/db/asm/x86_32:2167]: The primary reason for the change is to make the resulting machine code the same as the one produced by other assemblers, namely, NASM and FASM.

Technically, the original output was not incorrect as I stated. It is about which to choose when there are two options. Previously, when encoding `xchg r32, r32`, rasm2 chose the encoding of `xchg r/m32, r32`, whose `Op/En` is `MR`, which means that the left register operand is encoded in the `r/m` part of `ModR/M` and the right in the `reg` part. However, NASM and FASM choose `xchg r32, r/m32` when encoding `xchg r32, r32`. I thought it was a good idea to follow the assemblers widely in use.

The code making the choice is as follows. Flipping 0 and 1 would undo the change, which I can do if desired.

https://github.com/atodekangae/radare2/blob/2f5b8f1d672de7f1933c865293b51e0aac83222d/libr/asm/p/asm_x86_nz.c#L2791-2792

Comment (ret2libc) [reply test/db/asm/x86_32:2167]: @radare opinions?

Comment (radare) [reply test/db/asm/x86_32:2167]: uhm the swap of arguments in test and exchagne instructions on x86 assemnblers and disassemblers have been always a known issue. the thing is.. this change seems to return a different output than capstone, 
``
$ rasm2 -a x86  -d 87cd
xchg ebp, ecx
```

File: test/db/asm/x86_64 [hunk 2]
```diff
@@ -956,6 +956,24 @@ a "dec rdi" 48ffcf
...
+a "xchg rax, r8" 4990
+a "xchg r8, rax" 4990
+a "xchg rdx, rbx" 4887d3
+a "xchg rbx, rdx" 4887da
+a "xchg r8, r15" 4d87c7
+a "xchg r15, r8" 4d87f8
+a "xchg r8d, r15d" 4587c7
+a "xchg r15d, r8d" 4587f8
+a "xchg rdx, r8" 4987d0
+a "xchg r15, rdx" 4c87fa
```
Comment (radare) [inline test/db/asm/x86_64:976]: looks wrong to me:
```
$  rasm2 -a x86 -b64  -d 4c87fa
xchg rdx, r15
```

**Comment by atodekangae** [comment]
Thank you for the feedback. I made a change so that encoding is consistent with Capstone and it roundtips.

**Comment by atodekangae** [comment]
Thank you for the feedback. I made a change so that encoding is consistent with Capstone and it roundtrips.

**Comment by radare** [comment]
Looks good now! thanks!

File: test/db/asm/x86_64 [hunk 3]
```diff
@@ -956,6 +956,24 @@ a "dec rdi" 48ffcf
...
+a "bswap rax" 480fc8
+a "bswap r15" 490fcf
+a "bswap eax" 0fc8
+a "bswap r15d" 410fcf
+ad "xchg eax, r8d" 4190
+a "xchg r8d, eax" 4190
+ad "xchg rax, rdx" 4892
+a "xchg rdx, rax" 4892
+ad "xchg rax, r8" 4990
+a "xchg r8, rax" 4990
```
Comment (radare) [inline test/db/asm/x86_64:968]: this doesnt roundtrip

Comment (atodekangae) [reply test/db/asm/x86_64:968]: `xchg rax, rdx` and `xchg rdx, rax` have exactly the same encoding, `4892`. So either of the two cannot roundtrip. The same goes for the pair `xchg rax, r8` and `xchg r8, rax`. In general, there is an encoding for the cases where one of the operands is `ax`, `eax`, or `rax`, where the order of the operands is lost. 

Following is an excerpt from the Instruction Set Reference Manual. As you see, the encoding does not retain the order if an `ax` register is present.

<table>
<tr>
<td>90+<em>rw</em></td>
<td>XCHG AX, <em>r16</em></td>
<td>O</td>
<td>Valid</td>
<td>Valid</td>
<td>Exchange <em>r16</em> with AX.</td></tr>
<tr>
<td>90+<em>rw</em></td>
<td>XCHG <em>r16</em>, AX</td>
<td>O</td>
<td>Valid</td>
<td>Valid</td>
<td>Exchange AX with <em>r16.</em></td></tr>
<tr>
<td>90+<em>rd</em></td>
<td>XCHG EAX, <em>r32</em></td>
<td>O</td>
<td>Valid</td>
<td>Valid</td>
<td>Exchange <em>r32</em> with EAX.</td></tr>
<tr>
<td>REX.W + 90+<em>rd</em></td>
<td>XCHG RAX, <em>r64</em></td>
<td>O</td>
<td>Valid</td>
<td>N.E.</td>
<td>Exchange <em>r64</em> with RAX.</td></tr>
<tr>
<td>90+<em>rd</em></td>
<td>XCHG <em>r32</em>, EAX</td>
<td>O</td>
<td>Valid</td>
<td>Valid</td>
<td>Exchange EAX with <em>r32.</em></td></tr>
<tr>
<td>REX.W + 90+<em>rd</em></td>
<td>XCHG <em>r64</em>, RAX</td>
<td>O</td>
<td>Valid</td>
<td>N.E.</td>
<td>Exchange RAX with <em>r64.</em></td></tr>
</table>

Comment (atodekangae) [reply test/db/asm/x86_64:968]: As a side note, I noticed there is a test present to check if `xchg eax, eax` assembles to `90`. This does not roundtrip, but I kept the original behavior and test unchanged.

**Comment by radare** [comment]
👍👍

**Pull request merged**